aten/src/ATen/native/SegmentReduce.cpp

```diff
@@ -34,15 +35,22 @@ Tensor _segment_reduce_cpu_kernel(
               : std::numeric_limits<scalar_t>::lowest();
           for (int64_t j = 0; j < lengths_data[i]; ++j) {
             const auto data = values_data[k];
-            initial_value = at::_isnan(data)
-                ? data
-                : std::max<scalar_t>(initial_value, data);
+            // TODO: There is no need to branch with every element
+            if (reduction == SegmentReductionType::MAX) {
+              initial_value = at::_isnan(data)
+                  ? data
+                  : std::max<scalar_t>(initial_value, data);
+            } else if (reduction == SegmentReductionType::MEAN) {
+              initial_value = at::_isnan(data) ? data : (initial_value + data);
+            }
             k++;
           }
-          // If unsafe is false, check on lengths or indices should cover cases
-          // where lengths for a particular segment is negative. If unsafe
-          // is true, simply set to initial_value for particular reduction
+          TORCH_CHECK(lengths_data[i] >= 0);
           output_data[i] = initial_value;
+          if (reduction == SegmentReductionType::MEAN && lengths_data[i] > 0 &&
+              !at::_isnan(output_data[i])) {
+            output_data[i] = output_data[i] / lengths_data[i];
```

Comment:
This is a different behavior from regular `torch.mean`. `torch.mean` produces `nan` when averaging over 0 elements (whereas you would produce 0)

Reply:
Yep, that sounds good! (That's the same we do for max, right?)